src/translator/translator.h

```diff
@@ -123,18 +126,48 @@ class Translate : public ModelTask {
     data::BatchGenerator<data::Corpus> bg(corpus_, options_);
 
     ThreadPool threadPool(numDevices_, numDevices_);
-
+  
     size_t batchId = 0;
     auto collector = New<OutputCollector>(options_->get<std::string>("output"));
     auto printer = New<OutputPrinter>(options_, trgVocab_);
     if(options_->get<bool>("quiet-translation"))
       collector->setPrintingStrategy(New<QuietPrinting>());
 
-    bg.prepare();
+    // mutex for syncing counter and timer updates
+    std::mutex syncCounts;
+
+    // timer and counters for total elapsed time and statistics
+    std::unique_ptr<timer::Timer> totTimer(new timer::Timer()); 
+    size_t totBatches      = 0;
+    size_t totLines        = 0;
+    size_t totSourceTokens = 0;
+    
+    // timer and counters for elapsed time and statistics between updates
+    std::unique_ptr<timer::Timer> curTimer(new timer::Timer());
+    size_t curBatches      = 0;
+    size_t curLines        = 0;
+    size_t curSourceTokens = 0;
+
+    // determine if we want to display timer statistics, by default off
+    auto statFreq = SchedulingParameter::parse(options_->get<std::string>("stat-freq", "0u"));
+    ABORT_IF(statFreq.unit != SchedulingUnit::updates, "Units other than 'u' are not supported for --stat-freq value {}", statFreq);
```

Comment:
This is checked before batches generation, in particular before (potentially very long) shuffling, right? Maybe left a comment that it's checked before `bg.prepare()` because we want to notify users about a potential value error as soon as possible.

Reply:
Good point. Didn't really think about that TBH :)